llvm/include/llvm/Target/TargetMachine.h

```diff
@@ -531,6 +532,39 @@ class LLVM_ABI TargetMachine {
   }
 };
 
+using PassConfigCallback =
+    std::function<void(TargetMachine &, PassManagerBase &, TargetPassConfig *)>;
+
+class TargetPassConfigCallbackRegistry : public PassConfigCallback {
+protected:
+  static SmallVector<PassConfigCallback *, 1> Callbacks;
+
+  explicit TargetPassConfigCallbackRegistry(PassConfigCallback &&C)
+      : PassConfigCallback(std::move(C)) {}
+
+public:
+  static void invokeCallbacks(TargetMachine &TM, PassManagerBase &PM,
+                              TargetPassConfig *PassConfig) {
+    for (const PassConfigCallback *C : Callbacks)
+      (*C)(TM, PM, PassConfig);
+  }
+};
+
+class RegisterTargetPassConfigCallback
+    : public TargetPassConfigCallbackRegistry {
```

Comment:
This is just a RAII wrapper right? Let's make the callback a member and not inherit from std::function